table/metadata.go

```diff
@@ -80,20 +86,512 @@ type Metadata interface {
+		updates:       make([]Update, 0),
+		schemaList:    make([]*iceberg.Schema, 0),
+		specs:         make([]iceberg.PartitionSpec, 0),
+		props:         make(iceberg.Properties),
+		snapshotList:  make([]Snapshot, 0),
+		snapshotLog:   make([]SnapshotLogEntry, 0),
+		metadataLog:   make([]MetadataLogEntry, 0),
+		sortOrderList: make([]SortOrder, 0),
+		refs:          make(map[string]SnapshotRef),
+	}, nil
+}
+
+func MetadataBuilderFromBase(metadata Metadata) (*MetadataBuilder, error) {
+	b := &MetadataBuilder{}
+	b.base = metadata
+
+	b.formatVersion = metadata.Version()
+	b.uuid = metadata.TableUUID()
+	b.loc = metadata.Location()
+	b.lastUpdatedMS = metadata.LastUpdatedMillis()
+	b.lastColumnId = metadata.LastColumnID()
+	b.schemaList = metadata.Schemas()
+	b.currentSchemaID = metadata.CurrentSchema().ID
+	b.specs = metadata.PartitionSpecs()
+	b.defaultSpecID = metadata.DefaultPartitionSpec()
+	b.lastPartitionID = metadata.LastPartitionSpecID()
+	b.props = metadata.Properties()
+	b.snapshotList = metadata.Snapshots()
+	b.currentSnapshotID = &metadata.CurrentSnapshot().SnapshotID
+	b.sortOrderList = metadata.SortOrders()
+	b.defaultSortOrderID = metadata.DefaultSortOrder()
+	b.refs = metadata.Refs()
+	b.snapshotLog = metadata.SnapshotLogs()
+	b.metadataLog = metadata.PreviousFiles()
+
+	return b, nil
+}
+
+func (b *MetadataBuilder) AddSchema(schema *iceberg.Schema, newLastColumnID int) (*MetadataBuilder, error) {
+	if newLastColumnID < b.lastColumnId {
+		return nil, fmt.Errorf("invalid last column id %d, must be >= %d",
+			newLastColumnID, b.lastColumnId)
+	}
+
+	b.updates = append(b.updates, NewAddSchemaUpdate(schema, newLastColumnID))
+	b.lastColumnId = newLastColumnID
+	b.schemaList = append(b.schemaList, schema)
+
+	return b, nil
+}
+
+func (b *MetadataBuilder) AddPartitionSpec(spec *iceberg.PartitionSpec, initial bool) (*MetadataBuilder, error) {
+	for _, s := range b.specs {
+		if s.ID() == spec.ID() && !initial {
+			return nil, fmt.Errorf("partition spec with id %d already exists", spec.ID())
+		}
+	}
+
+	var maxFieldID int
+	if len(spec.Fields()) > 0 {
+		maxField := slices.MaxFunc(spec.Fields(), func(a, b iceberg.PartitionField) int {
+			return a.FieldID - b.FieldID
+		})
+		maxFieldID = maxField.FieldID
+	}
+
+	prev := PARTITION_FIELD_ID_START - 1
+	if b.lastPartitionID != nil {
+		prev = *b.lastPartitionID
+	}
+
+	lastPartitionID := max(maxFieldID, prev)
+	b.lastPartitionID = &lastPartitionID
+	b.specs = append(b.specs, *spec)
+	b.updates = append(b.updates, NewAddPartitionSpecUpdate(spec, initial))
+
+	return b, nil
+}
+
+func (b *MetadataBuilder) AddSnapshot(snapshot *Snapshot) (*MetadataBuilder, error) {
+	if len(b.schemaList) == 0 {
+		return nil, errors.New("can't add snapshot with no added schemas")
+	} else if len(b.specs) == 0 {
+		return nil, errors.New("can't add snapshot with no added partition specs")
+	} else if len(b.sortOrderList) == 0 {
+		return nil, errors.New("can't add snapshot with no added sort orders")
```

Comment:
why can't a snapshot exist without any sort order?

Reply:
If no sort order should be used, then the UnsortedSortOrder should be set if I understand correctly? So we check that the sortOrder has been specified.

Reply:
`UnsortedSortOrder` is the default returned if there's no sort order set. thus you can have an empty `sortOrderList` and `UnsortedSortOrder` will be returned as the default. there's no reason we shouldn't allow a snapshot with no sort order.